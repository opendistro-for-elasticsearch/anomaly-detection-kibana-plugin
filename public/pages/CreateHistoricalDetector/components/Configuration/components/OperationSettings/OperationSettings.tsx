/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import {
  EuiCheckbox,
  EuiText,
  EuiFlexItem,
  EuiFlexGroup,
  EuiFieldNumber,
} from '@elastic/eui';
import { Field, FieldProps, FormikProps } from 'formik';
import React from 'react';
import { FormattedFormRow } from '../../../../../createDetector/components/FormattedFormRow/FormattedFormRow';
import ContentPanel from '../../../../../../components/ContentPanel/ContentPanel';
import {
  getError,
  isInvalid,
  validatePositiveInteger,
} from '../../../../../../utils/utils';
import {
  HistoricalDetectorFormikValues,
  SAVE_HISTORICAL_DETECTOR_OPTIONS,
} from '../../../../utils/constants';

interface OperationSettingsProps {
  formikProps: FormikProps<HistoricalDetectorFormikValues>;
  selectedOption: SAVE_HISTORICAL_DETECTOR_OPTIONS;
  onOptionChange(id: string): void;
}

export function OperationSettings(props: OperationSettingsProps) {
  return (
    <ContentPanel title="Operation settings" titleSize="s">
      <EuiFlexGroup direction="column" style={{ margin: '0px' }}>
        <Field name="detectionInterval" validate={validatePositiveInteger}>
          {({ field, form }: FieldProps) => (
            <EuiFlexGroup direction="column">
              <EuiFlexItem style={{ maxWidth: '70%' }}>
                <FormattedFormRow
                  fullWidth
                  title="Detection interval"
                  hint="Define how often the detector collects data to generate
                    anomalies. The shorter the interval is, the more
                    detector results there will be, and the more computing resources
                    the detector will need."
                  isInvalid={isInvalid(field.name, form)}
                  error={getError(field.name, form)}
                >
                  <EuiFlexGroup gutterSize="s" alignItems="center">
                    <EuiFlexItem grow={false}>
                      <EuiFieldNumber
                        name="detectionInterval"
                        id="detectionInterval"
                        placeholder="Detector interval"
                        data-test-subj="detectionInterval"
                        {...field}
                      />
                    </EuiFlexItem>
                    <EuiFlexItem>
                      <EuiText>
                        <p className="minutes">minutes</p>
                      </EuiText>
                    </EuiFlexItem>
                  </EuiFlexGroup>
                </FormattedFormRow>
              </EuiFlexItem>
              <EuiFlexItem>
                <EuiCheckbox
                  id={'runDetectorAutomaticallyCheckbox'}
                  label="Run detector automatically after creating"
                  checked={
                    props.selectedOption ===
                    SAVE_HISTORICAL_DETECTOR_OPTIONS.START
                  }
                  disabled={false}
                  onChange={() => {
                    const reverseOption =
                      props.selectedOption ===
                      SAVE_HISTORICAL_DETECTOR_OPTIONS.START
                        ? SAVE_HISTORICAL_DETECTOR_OPTIONS.KEEP_STOPPED
                        : SAVE_HISTORICAL_DETECTOR_OPTIONS.START;
                    props.onOptionChange(reverseOption);
                  }}
                />
              </EuiFlexItem>
            </EuiFlexGroup>
          )}
        </Field>
      </EuiFlexGroup>
    </ContentPanel>
  );
}
