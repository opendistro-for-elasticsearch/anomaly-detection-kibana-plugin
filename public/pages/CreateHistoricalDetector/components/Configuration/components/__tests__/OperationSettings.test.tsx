/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React from 'react';
import { render, fireEvent, wait } from '@testing-library/react';
import { OperationSettings } from '../OperationSettings/OperationSettings';
import { Formik } from 'formik';
import userEvent from '@testing-library/user-event';

const TITLE_TEXT = 'Operation settings';
const INTERVAL_TEXT = 'Detection interval';
const RUN_DETECTOR_TEXT = 'Run detector automatically after creating';
const INVALID_TEXT = 'Must be a positive integer';

const onOptionChange = jest.fn();

const renderWithFormik = () => ({
  ...render(
    <Formik initialValues={{}} onSubmit={jest.fn()}>
      {() => (
        <div>
          <OperationSettings
            formikProps={{
              setFieldValue: jest.fn(),
              setFieldTouched: jest.fn(),
            }}
            onOptionChange={onOptionChange}
          />
        </div>
      )}
    </Formik>
  ),
});

describe('<OperationSettings /> spec', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });
  test('renders the component', () => {
    const { container, getByText, queryByText } = renderWithFormik();
    expect(container.firstChild).toMatchSnapshot();
    getByText(TITLE_TEXT);
    getByText(INTERVAL_TEXT);
    getByText(RUN_DETECTOR_TEXT);
    expect(queryByText(INVALID_TEXT)).toBeNull();
  });
  test('allows changing to another valid number', async () => {
    const { getByText, queryByText, getByPlaceholderText } = renderWithFormik();
    userEvent.type(getByPlaceholderText('Detector interval'), '1');
    fireEvent.blur(getByPlaceholderText('Detector interval'));
    await wait();
    getByText(TITLE_TEXT);
    expect(queryByText(INVALID_TEXT)).toBeNull();
  });
  test('catches invalid interval of 0', async () => {
    const { getByText, getByPlaceholderText } = renderWithFormik();
    userEvent.type(getByPlaceholderText('Detector interval'), '0');
    fireEvent.blur(getByPlaceholderText('Detector interval'));
    await wait();
    getByText(TITLE_TEXT);
    getByText(INVALID_TEXT);
  });
  test('catches invalid interval of -1', async () => {
    const { getByText, getByPlaceholderText } = renderWithFormik();
    userEvent.type(getByPlaceholderText('Detector interval'), '-1');
    fireEvent.blur(getByPlaceholderText('Detector interval'));
    await wait();
    getByText(TITLE_TEXT);
    getByText(INVALID_TEXT);
  });
  test('catches invalid interval that is non-numeric', async () => {
    const { getByText, getByPlaceholderText } = renderWithFormik();
    userEvent.type(getByPlaceholderText('Detector interval'), 'test');
    fireEvent.blur(getByPlaceholderText('Detector interval'));
    await wait();
    getByText(TITLE_TEXT);
    getByText(INVALID_TEXT);
  });
  test('changing the run option calls onOptionChange', () => {
    const { getByText } = renderWithFormik();
    expect(onOptionChange).toHaveBeenCalledTimes(0);
    fireEvent.click(getByText('Run detector automatically after creating'));
    expect(onOptionChange).toHaveBeenCalledTimes(1);
  });
});
