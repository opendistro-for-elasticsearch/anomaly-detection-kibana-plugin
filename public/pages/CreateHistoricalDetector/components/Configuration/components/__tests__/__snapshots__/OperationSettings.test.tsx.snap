// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OperationSettings /> spec renders the component 1`] = `
<div>
  <div
    class="euiPanel euiPanel--paddingMedium"
    style="padding: 20px;"
  >
    <div
      class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
      style="padding: 0px;"
    >
      <div
        class="euiFlexItem"
      >
        <h3
          class="euiTitle euiTitle--small"
        >
          Operation settings
        </h3>
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem"
          />
        </div>
      </div>
      <div
        class="euiFlexItem euiFlexItem--flexGrowZero"
      >
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem"
          />
        </div>
      </div>
    </div>
    <hr
      class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginXSmall"
    />
    <div
      style="padding: 10px 0px;"
    >
      <div
        class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionColumn euiFlexGroup--responsive"
        style="margin: 0px;"
      >
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionColumn euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem"
            style="max-width: 70%;"
          >
            <div
              class="euiFormRow euiFormRow--fullWidth"
              hint="Define how often the detector collects data to generate anomalies. Shorter intervals generate more detector results and  therefore require more computing resources."
              id="random_html_id-row"
              title="Detection interval"
            >
              <div
                class="euiFormRow__labelWrapper"
              >
                <label
                  aria-invalid="false"
                  class="euiFormLabel euiFormRow__label"
                  for="random_html_id"
                >
                  <div
                    style="line-height: 8px;"
                  >
                    <p>
                      Detection interval
                    </p>
                    <br />
                    <p
                      class="sublabel"
                    >
                      Define how often the detector collects data to generate anomalies. Shorter intervals generate more detector results and  therefore require more computing resources.
                    </p>
                  </div>
                </label>
              </div>
              <div
                class="euiFormRow__fieldWrapper"
              >
                <div
                  class="euiFlexGroup euiFlexGroup--gutterSmall euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow euiFlexGroup--responsive"
                  id="random_html_id"
                >
                  <div
                    class="euiFlexItem euiFlexItem--flexGrowZero"
                  >
                    <div
                      class="euiFormControlLayout"
                    >
                      <div
                        class="euiFormControlLayout__childrenWrapper"
                      >
                        <input
                          class="euiFieldNumber"
                          data-test-subj="detectionInterval"
                          id="detectionInterval"
                          name="detectionInterval"
                          placeholder="Detector interval"
                          type="number"
                          value=""
                        />
                      </div>
                    </div>
                  </div>
                  <div
                    class="euiFlexItem"
                  >
                    <div
                      class="euiText euiText--medium"
                    >
                      <p
                        class="minutes"
                      >
                        minutes
                      </p>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="euiFlexItem"
          >
            <div
              class="euiCheckbox"
            >
              <input
                class="euiCheckbox__input"
                id="runDetectorAutomaticallyCheckbox"
                type="checkbox"
              />
              <div
                class="euiCheckbox__square"
              />
              <label
                class="euiCheckbox__label"
                for="runDetectorAutomaticallyCheckbox"
              >
                Run detector automatically after creating
              </label>
            </div>
          </div>
          <div
            class="euiFlexItem"
            style="margin-top: -8px;"
          >
            <p
              class="sublabel"
            >
               
              We recommend running this detector during low traffic times.
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
